File: src/main/java/net/gtaun/shoebill/amx/AmxCallableImpl.java
package net.gtaun.shoebill.amx;

import net.gtaun.shoebill.SampNativeFunction;
import net.gtaun.shoebill.amx.types.ReturnType;

/**
 * Created by marvin on 20.12.2014 in project shoebill-runtime.
 * Copyright (c) 2014 Marvin Haschker. All rights reserved.
 */
public class AmxCallableImpl implements AmxCallable {

    private AmxInstance amxInstance;
    private String name;
    private AmxCallableType type;
    private int id;
    private ReturnType returnType;

    public AmxCallableImpl(AmxInstance instance, int id, String name, AmxCallableType type, ReturnType returnType) {
        this.amxInstance = instance;
        this.name = name;
        this.type = type;
        this.id = id;
        this.returnType = returnType;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public ReturnType getReturnType() {
        return returnType;
    }

    @Override
    public Object call(Object... args) {
        for(int i = 0; i < args.length; i++) {
            if(args[i] instanceof Boolean) {
                Boolean bool = (Boolean) args[i];
                args[i] = (bool) ? 1 : 0;
            }
        }
        try {
            if (type == AmxCallableType.NATIVE)
                return SampNativeFunction.callFunction(amxInstance.getHandle(), id, returnType.getValue(), args);
            else if (type == AmxCallableType.PUBLIC)
                return SampNativeFunction.callPublic(amxInstance.getHandle(), id, returnType.getValue(), args);
        } catch (Throwable e) {
            e.printStackTrace();
            return null;
        }
        return null;
    }
}

File: src/main/java/net/gtaun/shoebill/amx/AmxCallableImpl.kt
package net.gtaun.shoebill.amx

import net.gtaun.shoebill.SampNativeFunction
import net.gtaun.shoebill.amx.types.ReturnType
import org.slf4j.LoggerFactory

/**
 * @author Marvin Haschker
 */
internal class AmxCallableImpl(private val amxInstance: AmxInstance, private val id: Int,
                               override val name: String, private val type: AmxCallableType,
                               override val returnType: ReturnType) : AmxCallable() {

    override fun call(vararg args: Any): Any? {
        val arguments: Array<Any> = arrayOf(*args)
        args.filter { it is Boolean }.forEachIndexed { index, any ->
            arguments[index] = if((any as Boolean) == true) 1 else 0
        }

        arguments.forEachIndexed { index, any ->
            arguments[index] = AmxCallableClasses.primitiveToComposite(any)
        }

        val types: MutableList<String> = mutableListOf()
        val VALID_CLASSES = AmxCallableClasses.VALID_CLASSES

        arguments.forEach {
            if (!VALID_CLASSES.containsKey(it.javaClass)) {
                LOGGER.error("Cannot convert ${it.javaClass.simpleName} to a basic type. Please make sure that you " +
                        "use a valid class for the usage of the native interface.")
                return null
            } else {
                types.add(VALID_CLASSES[it.javaClass]!!)
            }
        }

        try {
            if (type == AmxCallableType.NATIVE)
                return SampNativeFunction.callFunction(amxInstance.handle, id, returnType.value, arguments, *types.toTypedArray())
            else if (type == AmxCallableType.PUBLIC)
                return SampNativeFunction.callPublic(amxInstance.handle, id, returnType.value, arguments, *types.toTypedArray())
        } catch (e: Throwable) {
            e.printStackTrace()
        }
        return null
    }

    companion object {

        val LOGGER = LoggerFactory.getLogger("Amx Callable")

    }
}
